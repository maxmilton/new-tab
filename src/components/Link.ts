import { h, S1Node } from 'stage1';

export interface LinkProps {
  title: string;
  url: string;
}

export type LinkComponent = S1Node & HTMLAnchorElement;

type RefNodes = {
  i: HTMLImageElement;
  t: Text;
};

const view = h(`
  <a>
    <img decoding=async #i>
    #t
  </a>
`);

export const Link = (item: LinkProps): LinkComponent => {
  const root = view.cloneNode(true) as LinkComponent;
  const nodes = view.collect<RefNodes>(root);

  root.href = item.url;
  root.title = item.title;
  nodes.i.src = 'chrome://favicon/' + item.url;
  nodes.t.nodeValue = item.title;

  return root;
};
