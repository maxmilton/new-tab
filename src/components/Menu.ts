import { h, type S1Node } from 'stage1';

type MenuComponent = S1Node & HTMLDivElement;
type RefNodes = {
  s: HTMLAnchorElement;
};

// https://github.com/tailwindlabs/heroicons/blob/master/optimized/outline/menu.svg
const view = h(`
  <div id=m>
    <svg id=im>
      <path d="M4 6h16M4 12h16M4 18h16" />
    </svg>

    <div id=d>
      <a href=chrome://new-tab-page>Open Default Tab</a>
      <a href=chrome://bookmarks>Bookmarks Manager</a>
      <a href=chrome://downloads>Downloads</a>
      <a href=chrome://history>History</a>
      <a href=chrome://settings/passwords>Passwords</a>

      <hr>

      <a #s>New Tab Settings</a>
      <a href=https://github.com/maxmilton/new-tab/issues>Report Bug</a>
    </div>
  </div>
`);

export const Menu = (): MenuComponent => {
  const root = view as MenuComponent;
  const nodes = view.collect<RefNodes>(root);

  nodes.s.__click = () => chrome.runtime.openOptionsPage();

  return root;
};
