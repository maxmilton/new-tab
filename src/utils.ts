// NOTE: When updating also update references that get an item by index
export const DEFAULT_SECTION_ORDER = [
  'Open Tabs',
  'Bookmarks',
  'History',
  'Top Sites',
  'Recently Closed Tabs',
];

export const handleClick = (event: MouseEvent): void => {
  const { target, ctrlKey } = event;
  const url = (target as HTMLAnchorElement).href;

  // only apply special handling to non-http links
  if (url && !url.startsWith('h')) {
    event.preventDefault();

    if ((target as HTMLAnchorElement).target === '_blank' || ctrlKey) {
      // open the location in a new tab
      void chrome.tabs.create({ url });
    } else {
      // update the location in the current tab
      void chrome.tabs.update({ url });
    }
  }

  // focus search element on bad click
  if(target.tagName=== 'BODY') {
    document.getElementById('s').focus();
  }
};
