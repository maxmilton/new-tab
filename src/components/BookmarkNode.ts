/* eslint-disable @typescript-eslint/restrict-plus-operands */

import { append, create, h, type S1Node } from 'stage1';
import { Link, type LinkComponent, type LinkProps } from './Link';

type FolderPopupComponent = HTMLDivElement & {
  adjustPosition(): void;
};

const CLOSE_DELAY_MS = 600;
let emptyPopupView: S1Node | undefined;
let arrowView: S1Node | undefined;

const folderPopupView = create('div');
folderPopupView.className = 'sf';

const FolderPopup = (
  parent: Element,
  children: chrome.bookmarks.BookmarkTreeNode[],
  topLevel: boolean,
): FolderPopupComponent => {
  const root = folderPopupView.cloneNode(true) as FolderPopupComponent;
  const parentRect = parent.getBoundingClientRect();
  let top: number;
  let left: number;

  if (topLevel) {
    // Show top level folder popup bellow its parent
    top = parentRect.bottom;
    left = parentRect.left;
  } else {
    // Show nested folder popup beside its parent
    top = parentRect.top;
    left = parentRect.right;
  }

  root.style.cssText = `top:${top}px;left:${left}px;max-height:${
    window.innerHeight - top
  }px`;

  if (children.length) {
    children.forEach((item) => {
      append(BookmarkNode(item), root);
    });
  } else {
    append((emptyPopupView ??= h('<div id=e>(empty)</div>')), root);
  }

  // Only after the component is mounted in the DOM do we have element size
  // information so final position adjustment is split into a separate step
  root.adjustPosition = () => {
    const viewportWidth = document.documentElement.clientWidth;
    const width = root.offsetWidth;

    if (left + width > viewportWidth) {
      // Show top level aligned to the right edge of the viewport
      // Show nested show to the left of its parent
      root.style.left = topLevel
        ? viewportWidth - width + 'px'
        : parentRect.left - width + 'px';
    }
  };

  return root;
};

type FolderComponent = HTMLDivElement & {
  closePopup(this: void): void;
};

export interface FolderProps
  extends Omit<chrome.bookmarks.BookmarkTreeNode, 'id'> {
  end?: boolean;
}

const folderView = create('div');
folderView.className = 'f';

export const Folder = (props: FolderProps): FolderComponent => {
  const root = folderView.cloneNode(true) as FolderComponent;
  let popup: FolderPopupComponent | null;
  let timer: NodeJS.Timeout;

  const clearTimer = () => clearTimeout(timer);

  const resetTimer = () => {
    clearTimer();
    timer = setTimeout(root.closePopup, CLOSE_DELAY_MS);
  };

  root.textContent = props.title;

  if (props.end) root.className += ' end';

  // parentId 0 = "bookmarks bar", 1 = "other bookmarks"
  if (props.parentId! > '1') {
    append(
      (arrowView ??= h(`
        <svg class=i>
          <path d="M5 12h14M12 5l7 7-7 7"/>
        </svg>
      `)).cloneNode(true),
      root,
    );
  }

  root.closePopup = () => {
    if (popup) {
      popup.remove();
      popup = null;
    }
  };

  root.__mouseover = () => {
    clearTimer();

    if (!popup) {
      const parent = root.parentNode as Element;

      // Immediately close any folder popups on the parent level
      parent
        .querySelectorAll<FolderComponent>('.f')
        .forEach((folder) => folder.closePopup());

      popup = FolderPopup(
        root,
        props.children!,
        // BookmarkBar (bookmarks top level) has attribute id="b"
        parent.id === 'b',
      );

      popup.__mouseover = clearTimer;
      popup.__mouseout = resetTimer;

      append(popup, root);
      popup.adjustPosition();
    }
  };

  root.__mouseout = resetTimer;

  return root;
};

export const BookmarkNode = (
  props: LinkProps | FolderProps,
  // @ts-expect-error - FIXME:!
): LinkComponent | FolderComponent => (props.url ? Link : Folder)(props);
