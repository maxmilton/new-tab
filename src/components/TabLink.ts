import { Link, type LinkComponent, type LinkProps } from './Link';

export interface TabLinkProps extends LinkProps {
  /** Tab ID. */
  id: number;
  windowId: number;
}

export const TabLink = (props: TabLinkProps): LinkComponent => {
  const root = Link(props);

  root.__click = () => {
    // Switch to the clicked tab
    void chrome.tabs.update(props.id, { active: true });

    // Switch active window if the tab isn't in the current window
    chrome.windows.getCurrent({}, (currentWindow) => {
      if (currentWindow.id !== props.windowId) {
        void chrome.windows.update(props.windowId, { focused: true });
      }
    });

    // Close current "new-tab" page
    chrome.tabs.getCurrent((currentTab) => {
      void chrome.tabs.remove(currentTab!.id!);
    });
  };

  return root;
};
