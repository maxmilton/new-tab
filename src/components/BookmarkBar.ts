import { append, create, h } from 'stage1';
import { BookmarkNode, Folder, type FolderProps } from './BookmarkNode';

declare global {
  interface HTMLElement {
    /**
     * BookmarkBar synthetic `mouseenter` event handler. Note the property is
     * named `__mouseover` but it actually works like `mouseenter`.
     */
    __mouseover(event: MouseEvent): void;
    /**
     * BookmarkBar synthetic `mouseleave` event handler. Note the property is
     * named `__mouseout` but it actually works like `mouseleave`.
     */
    __mouseout(event: MouseEvent): void;
  }
}

type BookmarkBarComponent = HTMLDivElement;

export const BookmarkBar = (): BookmarkBarComponent => {
  const root = create('div');
  root.id = 'b';

  chrome.bookmarks.getTree((tree) => {
    const [{ children: bookmarks }, otherBookmarks] = tree[0].children!;
    const len = bookmarks!.length;

    // Add BookmarkNodes one at a time until they can't fit in the bookmark
    // bar and then create a folder with the overflowing items
    const resize = () => {
      performance.mark('BookmarkBar');

      // Remove child nodes
      root.textContent = '';

      let otherBookmarksFolder;
      // NOTE: None of the elements we're measuring have a border or margin so
      // we can use clientWidth instead of offsetWidth for better performance
      let currentWidth = 0;

      if (otherBookmarks.children!.length) {
        (otherBookmarks as FolderProps).end = true;
        otherBookmarksFolder = append(Folder(otherBookmarks), root);
        currentWidth = otherBookmarksFolder.clientWidth;
      }

      if (len) {
        // minus overflow folder width (66 == 24px svg + 2*8px padding + 2*13px padding)
        const maxWidth = root.clientWidth - 66;
        let index = 0;

        for (; index < len; index++) {
          const node = append(BookmarkNode(bookmarks![index]), root);
          currentWidth += node.clientWidth;

          if (currentWidth >= maxWidth) {
            // There's no way to know an element's width before adding it to the
            // DOM so now we need to remove the added node which overflowed
            node.remove();
            break;
          }
        }

        if (index < len) {
          const overflowBookmarksFolder = append(
            Folder({
              // TODO: More elegant solution to show caret in overflow folder
              // children: bookmarks!.slice(index),
              children: bookmarks!.slice(index).map((item) => {
                // eslint-disable-next-line no-param-reassign
                item.parentId = '2';
                return item;
              }),
              end: true,
              title: '',
            }),
            root,
          );

          append(
            // https://github.com/feathericons/feather/blob/master/icons/corner-right-down.svg
            h(`
              <svg id=io>
                <polyline points="10 15 15 20 20 15"/>
                <path d="M4 4h7a4 4 0 0 1 4 4v12"/>
              </svg>
            `),
            overflowBookmarksFolder,
          );
        }

        // The "other bookmarks" folder was added first so overflow calculation
        // is correct but now move it to its proper position at the end
        if (otherBookmarksFolder) {
          append(otherBookmarksFolder, root);
        }
      }

      performance.measure('BookmarkBar', 'BookmarkBar');
    };

    // HACK: Because this script is loaded async, there exists a race condition
    // where the JS runs before the CSS is loaded. The styles are required to
    // calculate the width of each bookmark node. Loading the JS script async
    // yeilds the best load performance but a better overall solution is needed.
    const waitForStylesThenResize = () => {
      if (document.styleSheets.length) {
        resize();
      } else {
        setTimeout(waitForStylesThenResize);
      }
    };

    waitForStylesThenResize();

    window.onresize = resize;
  });

  // Synthetic `mouseenter` and `mouseleave` event handler
  // XXX: Similar to stage1 synthetic event logic but does not stop propagating
  // once an event handler is called + extra relatedTarget checks
  // https://github.com/maxmilton/stage1/blob/08cb3c08cb3e5513c181f768ae92c488cfe2a17a/src/events.ts#L3
  // eslint-disable-next-line no-multi-assign
  root.onmouseover = root.onmouseout = (event) => {
    const eventKey = ('__' + event.type) as '__mouseover' | '__mouseout';
    // null when mouse moves from/to outside the viewport
    const related = event.relatedTarget as Node | null;
    let node = event.target as
      | (Node & {
          __mouseover?(event2: MouseEvent): void;
          __mouseout?(event2: MouseEvent): void;
        })
      | null;

    while (node) {
      if (node[eventKey] && (!related || !node.contains(related))) {
        node[eventKey]!(event);
      }
      node = node.parentNode;
    }
  };

  return root;
};
