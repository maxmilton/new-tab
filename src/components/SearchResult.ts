import {
  append, createFragment, h, S1Node,
} from 'stage1';
import { DEFAULT_SECTION_ORDER } from '../utils';
import { Link, LinkProps } from './Link';
import { TabLink } from './TabLink';

export type SearchResultComponent = S1Node &
HTMLDivElement & {
  update: (this: void, newData: any[]) => void;
  filter: (this: void, text: string) => void;
};
type RefNodes = {
  t: Text;
  l: HTMLDivElement;
  m: HTMLButtonElement;
};

const DEFAULT_RESULTS_AMOUNT = 12; // chrome.topSites.get returns 12 items
const MORE_RESULTS_AMOUNT = 50;

const view = h(`
  <div hidden>
    <h2 #t></h2>

    <div #l></div>

    <button #m>Show more</button>
  </div>
`);

export const SearchResult = <T extends LinkProps>(
  sectionName: string,
): SearchResultComponent => {
  const root = view.cloneNode(true) as SearchResultComponent;
  const nodes = view.collect<RefNodes>(root);
  const isOpenTabs = sectionName === DEFAULT_SECTION_ORDER[0];
  let rawData: T[];
  let renderedLength: number;

  const renderList = (listData: T[], showCount = DEFAULT_RESULTS_AMOUNT) => {
    performance.mark(sectionName);

    const partial = isOpenTabs ? listData : listData.slice(0, showCount);
    const frag = createFragment();
    renderedLength = partial.length;

    partial.forEach((item) => {
      // @ts-expect-error - FIXME:!
      append((isOpenTabs ? TabLink : Link)(item), frag);
    });

    nodes.l.replaceChildren(frag);

    root.hidden = !renderedLength;
    nodes.m.hidden = renderedLength >= listData.length;

    performance.measure(sectionName, sectionName);
  };

  const update = (newData: T[]) => {
    renderList(newData);
    rawData = newData;
  };

  root.update = update;

  root.filter = (text) => renderList(
    rawData.filter(
      (item) => (item.title + '+' + item.url)
        .toLowerCase()
        .indexOf(text.toLowerCase()) > -1,
    ),
  );

  nodes.t.textContent = sectionName;

  nodes.m.__click = () => renderList(rawData, renderedLength + MORE_RESULTS_AMOUNT);

  return root;
};
