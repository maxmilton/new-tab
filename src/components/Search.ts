import { append, h, S1Node } from 'stage1';
import type { UserStorageData } from '../types';
import { DEFAULT_SECTION_ORDER } from '../utils';
import { SearchResult, SearchResultComponent } from './SearchResult';

type SectionRefs = Record<string, SearchResultComponent>;

const runSearch = (text: string, section: SectionRefs) => {
  const openTabs = section[DEFAULT_SECTION_ORDER[0]];
  const bookmarks = section[DEFAULT_SECTION_ORDER[1]];
  const history = section[DEFAULT_SECTION_ORDER[2]];
  const topSites = section[DEFAULT_SECTION_ORDER[3]];
  const recentlyClosed = section[DEFAULT_SECTION_ORDER[4]];

  if (history) {
    if (text) {
      chrome.history.search({ text }, history.update);
    } else {
      history.update([]);
    }
  }

  if (bookmarks) {
    if (text) {
      chrome.bookmarks.search(text, bookmarks.update);
    } else {
      bookmarks.update([]);
    }
  }

  openTabs?.filter(text);
  topSites?.filter(text);
  recentlyClosed?.filter(text);
};

type SearchComponent = S1Node & HTMLDivElement;
type RefNodes = {
  s: HTMLInputElement;
};

// https://github.com/feathericons/feather/blob/master/icons/search.svg
const view = h(`
  <div class=con>
    <input id=s placeholder="Search browser..." autocomplete=off #s>
    <svg id=i>
      <circle cx=11 cy=11 r=8 />
      <line x1=24 y1=24 x2=16.65 y2=16.65 />
    </svg>
  </div>
`);

export const Search = (): SearchComponent => {
  const root = view as SearchComponent;
  const { s } = view.collect<RefNodes>(root);
  const section: SectionRefs = {};

  s.oninput = () => runSearch(s.value, section);

  s.onkeyup = (event) => {
    if (event.key === 'Escape') {
      s.value = '';
      runSearch('', section);
    }
  };

  chrome.storage.local.get(null, (userSettings: UserStorageData) => {
    const sectionOrder = userSettings.o || DEFAULT_SECTION_ORDER;

    sectionOrder.forEach((name) => {
      section[name] = append(SearchResult(name), root);
    });

    const openTabs = section[DEFAULT_SECTION_ORDER[0]];
    const topSites = section[DEFAULT_SECTION_ORDER[3]];
    const recentlyClosed = section[DEFAULT_SECTION_ORDER[4]];

    if (openTabs) {
      const updateOpenTabs = () => chrome.tabs.query({}, (result) => {
        openTabs.update(result);
        if (s.value) openTabs.filter(s.value);
      });

      updateOpenTabs();

      chrome.tabs.onUpdated.addListener(updateOpenTabs);
      chrome.tabs.onRemoved.addListener(updateOpenTabs);
      chrome.tabs.onMoved.addListener(updateOpenTabs);

      // When opening multiple new-tab pages the browser will continue to update
      // the "Open Tabs" section on all pages, causing a significant performance
      // overhead. The impact is multiplied by the number of open tabs * the
      // number of new-tab pages. The actual problem is work is done even on
      // pages that are not visible and with execution on older pages before the
      // current page. Additionally, the SearchResult component is implemented
      // in a way that produces the smallest JS size and fast execution (remove
      // entire list DOM and insert new DOM fragment) rather than for efficiency
      // (e.g., DOM reconciliation; diff list DOM state and mutate changes
      // minimising adding or removing DOM nodes).

      // TODO: Keep? Causes significantly worse page load speed!

      // // When the page isn't active stop the "Open Tabs" section from updating to
      // // prevent performance issues when users open many new-tab pages.
      // document.onvisibilitychange = () => {
      //   if (document.hidden) {
      //     // @ts-expect-error - force override to kill API method
      //     chrome.tabs.query = () => {};
      //     // chrome.tabs.onUpdated.removeListener(updateOpenTabs);
      //     // chrome.tabs.onRemoved.removeListener(updateOpenTabs);
      //     // chrome.tabs.onMoved.removeListener(updateOpenTabs);
      //   } else {
      //     // eslint-disable-next-line no-restricted-globals
      //     location.reload();
      //   }
      // };
    }

    if (topSites) {
      chrome.topSites.get(topSites.update);
    }

    if (recentlyClosed) {
      chrome.sessions.getRecentlyClosed({}, (sessions) => {
        recentlyClosed.update(
          sessions.map((session) => session.tab).filter((tab) => tab),
        );
      });
    }
  });

  return root;
};
