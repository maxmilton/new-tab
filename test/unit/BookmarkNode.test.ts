import { afterEach, describe, expect, test } from 'bun:test';
import { BookmarkNode, type BookmarkTreeNode } from '../../src/components/BookmarkNode';
import { type LinkProps } from '../../src/components/Link';
import { cleanup, render } from './utils';

afterEach(cleanup);

test('renders a link when url is provided', () => {
  // @ts-expect-error - intentional partial props
  const rendered = render(BookmarkNode({ url: 'x' }));
  expect(rendered.container.querySelector('a')).toBeTruthy();
  expect(rendered.container.querySelector('div.f')).toBeFalsy();
});

test('renders a folder when url is not provided', () => {
  // @ts-expect-error - intentional partial props
  const rendered = render(BookmarkNode({}));
  expect(rendered.container.querySelector('div.f')).toBeTruthy();
  expect(rendered.container.querySelector('a')).toBeFalsy();
});

describe('Bookmark (Link)', () => {
  test('rendered DOM contains expected elements', () => {
    const rendered = render(
      BookmarkNode({ title: 'Example', url: 'https://example.com' } satisfies LinkProps),
    );
    const root = rendered.container.firstChild as HTMLElement;
    expect(root).toBeInstanceOf(window.HTMLAnchorElement);

    // TODO: More/better assertions to prove it's a Link component.
  });

  test('rendered DOM matches snapshot', () => {
    const rendered = render(BookmarkNode({ title: 'Example', url: 'https://example.com' }));
    expect(rendered.container.innerHTML).toMatchSnapshot();
  });

  // TODO: Consider running the same tests as for Link component.
  //  â†³ How might we do this elegantly without duplicating the tests?
});

describe('Folder', () => {
  test('rendered DOM contains expected elements', () => {
    const rendered = render(BookmarkNode({ id: '1', title: 'Example' } satisfies BookmarkTreeNode));
    const root = rendered.container.firstChild as HTMLElement;
    expect(root).toBeInstanceOf(window.HTMLDivElement);
    expect(root.className).toBe('f');
  });

  test('rendered DOM matches snapshot', () => {
    const rendered = render(BookmarkNode({ id: '1', title: 'Example' }));
    expect(rendered.container.innerHTML).toMatchSnapshot();
  });

  // TODO: Consider running the most tests twice, once for BookmarkNode and again for Folder directly (except for tests which use the Folder 3rd argument; "children").
  // TODO: Test BookmarkNode 2nd argument; "nested"
  // TODO: Test open/close state on top level
  // TODO: Test open/close state of sub folders ("popup")
  // TODO: Rendered DOM snapshot when popup is open
  // TODO: Test openning another folder closes the previous one if on the same or lower level
  // TODO: Test openned sub folder positions
  // TODO: Test openned sub folder is scrollable when overflowing (might not be possible with happy-dom, but should definately write an e2e test for this)
});
