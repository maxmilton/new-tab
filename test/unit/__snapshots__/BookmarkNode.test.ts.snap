// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark (Link) rendered DOM matches snapshot 1`] = `"<a href="https://example.com" title="Example"><img decoding="async" src="_favicon?size=16&amp;pageUrl=https%3A%2F%2Fexample.com">Example</a>"`;

exports[`Folder rendered DOM matches snapshot 1`] = `"<div class="f">Example</div>"`;
